source/common/stats/stat_merger.cc

```diff
@@ -0,0 +1,106 @@
+#include "common/stats/stat_merger.h"
+
+#include <regex>
+
+namespace Envoy {
+namespace Stats {
+
+StatMerger::StatMerger(Stats::Store& target_store) : target_store_(target_store) {}
+
+absl::optional<StatMerger::CombineLogic>
+StatMerger::getCombineLogic(const std::string& gauge_name) {
+  // Gauge name *substrings*, and special logic to use for combining those gauges' values.
+  static const std::vector<std::pair<std::regex, CombineLogic>> nonstandard_combine_logic{
+      // Any .version is either a static property of the binary, or an opaque identifier for
+      // resources that are not passed across hot restart.
+      {std::regex(".*\\.version$"), CombineLogic::NoImport},
```

Comment:
Thinking about this more; this is probably *not* running on the hot-path and only runs during a restart, so it's really not perf-critical. Feel free to leave this as is. It may be useful though to test how long this takes with (say) 10k clusters, generating 1M stats.

RE static-non-pod: this should be easy to resolve. I think we should use one of the lazy-init mechanisms to avoid the init-fiasco; see https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use

Reply:
Yes, it runs in the control thread, once per 5 seconds (by default), while a hot restart is happening.

I believe undefined init order is only an issue for static globals. This one is sitting in a function, which is itself the fix for that problem.

Reply:
I use resolve to signal that I'm not asking for a response. If I'm not sure how to proceed, I comment and leave unresolved, so that there will be the text box inviting a response. So, resolved is like "all good on this side"; the reviewer can always pick it back up.

I've been thinking it shouldn't be too bad because it's happening right alongside the stat flushing; there's already a loop over all gauges (and counters and histograms as well) happening. But, maybe these regex matches would be pretty bad; I'll add something to measure the performance to the tests.

Reply:
I think the protocol you are following isn't ideal for reviewers in github. Can I suggest:
 * if you did what the reviewer suggested, go ahead and resolve
 * if you want to push back, explain why and leave it unresolved, and let the reviewer resolve.

In a large PR it's difficult to find threads that are marked resolved, but that the reviewer feels still need addressing, without re-scanning the PR. Ie. "reviewer can always pick it back up" is not a good flow in github AFAICT.

Reply:
Ok, I took a look at the performance. With 100k gauge names, the current code takes ~10s. Switching back to the substrings I originally had brought it down to 4s. Maybe the trie approach I mentioned way back when would be better, although probably the amount of effort would be not much less than just going all the way to modifying the whole stat macro declaration interface.

One possible quicker fix might be to have the gauges carry their logic enum, but have it get set by consulting the logic exceptions table when the stat is created.

Reply:
There aren't very many gauges in practice so I don't think this is going to be a big deal. I would rather just ship and move directly to storing the merge type in the gauge itself which IMO is the real fix. If we assume that we are going to do that, I'm fine with moving back to the substring match if you prefer that for now. @jmarantz WDYT?

Reply:
Just to be clear: this is only used for gauges, not for counters, so it's not nearly as bad as it could be. Let's say there are 20 gauges per cluster; then we'd at 10k clusters we'd have 200k gauges which -- extrapolating from your data -- will take 20 seconds to run these regexes on. If we do that every 5 seconds it sounds like trouble.

Quick sanity check, Fred: you were using "-c opt" for your test, right?

Here are some localized hacks you could do:
 * switch to RE2 and get a dramatic speedup from a better library.  That may be too much
    infrastructural work for this though -- I think you'll like my second suggestion better :)
 * You have many spare bits in the flags in the stat structure now: use two of them to cache the result of the regex, so you only pay the regex crunching the first time you hit a particular stat. I think that hack could be very localized in this routine.

Having said all this, I think the team that would be most concerned about this performance hiccup, that uses both a large number of clusters and enables hot-restart is Istio. @PiotrSikora any thoughts on this?


On another note, WDYT of a trivial change to use a pointer to keep the static regex array -- that would be more consistent with the other construct-on-first-use idioms in Envoy and also be consistent with the recommendation of the "fiasco" document I pointed to above in this comment stream.

Reply:
What would be involved in just doing the work to store the merge type in the gauge right now? Can we just add new GAUGE macros for each merge type and force users to pick the merge type when they use a macro or directly instantiate a gauge? That doesn't seem that bad? The only issue I see with allowing the new process to set the gauge type is that we might not be able to do the "import if doesn't exist" type?

Reply:
Haha, oops, no I wasn't. With -c opt it's 0.6s for the substrings, 1.4s for regex.

I do think that hack sounds like a great idea, thanks. Best of all, with shared memory removed, it's no longer a big deal to make use of some of the flag bits: we can remove that use at any time without breaking hot restart compatibility.

That's an interesting point about the no-import type. Like, you could have a cluster that hasn't come up in the child yet, so its stats haven't been instantiated, so given just its name you don't have any knowledge of its combination logic? That does seem like a problem.

Reply:
> That does seem like a problem.

I don't think it's actually a problem, it's how it works today, meaning, stats don't show up in the new process until they "attach" to the existing shared memory, so IMO it's fine to drop this type completely now that I think about it.

Reply:
Given the performance with opt, I still think we should just ship this modulo some existing comments, and then circle back with the better gauge type solution as an independent change. WDYT?

Reply:
sgtm.